### DISCUSSION TITLE
Get "rests" with the musical symbol format (e.g., quarter rests)

### DISCUSSION THREAD

**Discussion by garcon33** [melanchall/drywetmidi#179] [Q&A] [answered]
Hello, 

Would anyone know if it's possible to get the length of "rests" with the musical symbol format (e.g., whole/half/quarter rests)?

I tried to do so with "GetObjects", but it seems to extract all the silent sections between notes rather than the rests described on music sheets.

Also, I'd like to know in which measure/bar the rests are.
Ideally, I'm trying to construct the data model like:

Array<track/channel> 
track/channel -> Array<measure (bar)> 
measure (bar) -> Array<notes/rests> (notes and rests are contained in order)

I'd be appreciated it if anyone could help with that. 

Many thanks!
Yj

**Comment by melanchall** [comment, marked as answer]
Hi,

First of all, there is no such term as "rest" in MIDI. More than that, there is no "note" term there. MIDI file contains **events**. Pair of Note On and Note Off events we call _a note_. Silence between such pairs we call _a rest_. What you see in a sheet music editor is not what will be stored in a MIDI file. MIDI file is just a set of instructions for a synth.

But your task looks pretty easy to solve:

```csharp
var tempoMap = midiFile.GetTempoMap();
var notesAndRests = midiFile
    .GetTrackChunks()
    .Select((trackChunk, i) => new
    {
        Index = i,
        Objects = trackChunk
            .GetObjects(
                ObjectType.Note | ObjectType.Rest,
                new ObjectDetectionSettings
                {
                    RestDetectionSettings = new RestDetectionSettings
                    {
                        RestSeparationPolicy = RestSeparationPolicy.NoSeparation
                    }
                })
            .Select(obj => new
            {
                Object = obj,
                Time = obj.TimeAs<BarBeatTicksTimeSpan>(tempoMap),
                Length = ((ILengthedObject)obj).LengthAs<MusicalTimeSpan>(tempoMap)
            })
            .ToArray()
    })
    .ToArray();
```

`notesAndRests` is an array of `A`, where `A` contains properties:
* `Index` – the index of a track chunk;
* `Objects` – an array of `B`, where `B` contains properties:  
  * `Object` – either [Note](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Interaction.Note.html) or [Rest](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Interaction.Rest.html);
  * `Time` – time of an object in terms of bars/beats, so you can easily know the bar where the object placed;
  * `Length` – length in terms of whole note fractions, so that's what you want, to know length like `1/4` or `1/2`.

Of course you can add other properties, modify parameters and so on. It's just a quick example of how you can get what you want. Please see the library documentation: https://melanchall.github.io/drywetmidi. Here articles on time spans: https://melanchall.github.io/drywetmidi/articles/high-level-managing/Time-and-length-overview.html.

**Comment by garcon33** [comment]
Hi melanchall,

Thanks a lot for your help!
It always helps me to understand the data structure of the library. I am new to working with MIDI data and music itself, so this is very helpful.

I understand that MIDI data is not meant for that purpose, but I wanted to extract as much information as possible described on a sheet music for my project...!

Best,
Yj
